### DISCUSSION TITLE
HeadObject error when trying to upload file to S3

### DISCUSSION THREAD

**Discussion by shlomi-viz** [apache/airflow#18803] [General] [answered]
### Apache Airflow Provider(s)

amazon

### Versions of Apache Airflow Providers

```apache-airflow-providers-amazon==1.4.0
apache-airflow-providers-ftp==2.0.1
apache-airflow-providers-http==1.1.1
apache-airflow-providers-imap==2.0.1
apache-airflow-providers-postgres==1.0.1
apache-airflow-providers-sftp==1.1.1
apache-airflow-providers-sqlite==2.0.1
apache-airflow-providers-ssh==1.3.0
```

### Apache Airflow version

2.0.2

### Operating System

Mac and Linux

### Deployment

MWAA

### What happened

When trying to upload a file to another AWS account using the `load_file` function, I got the following 
```
An error occurred (403) when calling the HeadObject operation: Forbidden
```
looking at source code the error is from 
```
    @provide_bucket_name
    @unify_bucket_name_and_key
    def check_for_key(self, key: str, bucket_name: Optional[str] = None) -> bool:
        """
        Checks if a key exists in a bucket

        :param key: S3 key that will point to the file
        :type key: str
        :param bucket_name: Name of the bucket in which the file is stored
        :type bucket_name: str
        :return: True if the key exists and False if not.
        :rtype: bool
        """
        try:
            self.get_conn().head_object(Bucket=bucket_name, Key=key)
            return True
        except ClientError as e:
            if e.response["ResponseMetadata"]["HTTPStatusCode"] == 404:
                return False
            else:
                raise e
```

As I understand it I'm getting this error since I don't a `List` permissions on the bucket in the other account

### What you expected to happen

AirFlow should try and upload the file, even if it already exists. In this section of the code we can add another check for the response code we get when no `List` permissions

```python
        except ClientError as e:
            if e.response["ResponseMetadata"]["HTTPStatusCode"] == 404:
                return False
            else:
                raise e
````

### How to reproduce

Use the following command to upload the file to an S3 bucket that you only have `Get` and `Put` permissions.
```python 
s3_hook.load_file(local_file_path, bucket_name=bucket_name, key=s3_key, acl_policy='bucket-owner-full-control')
````

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by eladkal** [comment]
> AirFlow should try and upload the file, even if it already exists

`load_file` has `replace` parameter (default False)
https://github.com/apache/airflow/blob/43f595fe1b8cd6f325d8535c03ee219edbf4a559/airflow/providers/amazon/aws/hooks/s3.py#L499-L500 
You don't set it in your example so it will not try to overwrite existed key.

Is your suggestion only catch the error from the API and change it to an Airflow error?
You are always welcome to open PR directly with your suggested code.

**Comment by shlomi-viz** [comment]
I want it to be able to upload a file in any case (file already exists or not), I know about the `replace` flag, but I also want to cover the case when the file doesn't exists. 

My suggestion is to catch the error and allow the upload, without adding another error. It will be something like, updating:
```python
        except ClientError as e:
            if e.response["ResponseMetadata"]["HTTPStatusCode"] == 404:
                return False
            else:
                raise e
```

to:

```python 
        except ClientError as e:
            if e.response["ResponseMetadata"]["HTTPStatusCode"] IN (404, 403):
                return False
            else:
                raise e
```

shortest PR ever 😆  I will try to submit one next week.

**Comment by shlomi-viz** [comment]
I created a PR https://github.com/apache/airflow/pull/18201, what is the process to get it reviewed? Thanks

**Comment by xinbinhuang** [comment]
> I want it to be able to upload a file in any case (file already exists or not), I know about the `replace` flag, but I also want to cover the case when the file doesn't exists.
> 
> My suggestion is to catch the error and allow the upload, without adding another error. It will be something like, updating:
> 
> ```python
>         except ClientError as e:
>             if e.response["ResponseMetadata"]["HTTPStatusCode"] == 404:
>                 return False
>             else:
>                 raise e
> ```
> 
> to:
> 
> ```python
>         except ClientError as e:
>             if e.response["ResponseMetadata"]["HTTPStatusCode"] IN (404, 403):
>                 return False
>             else:
>                 raise e
> ```
> 
> shortest PR ever 😆 I will try to submit one next week.

I believe the `replace=True` flag will just upload the file regardless if the key/file exists or not.  We don't call the `check_for_key` function when `replace=True`. Have you tried it?

IMO, It's not reasonable for us to swallow the permission error.

**Comment by shlomi-viz** [comment]
It will work with `replace=True`, but this miss-leading. If we don't want to `swallow the permission error` why do we check for the 404 error? 
I think that in both cases (with `List` permissions and without) the result should be the same.

**Reply by xinbinhuang** [reply]
They are 2 different things. The reason for checking `404` is because of the design of the S3 Head object API - it returns a 404 code when the object doesn't exist. (more info [here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.head_object))

**Reply by shlomi-viz** [reply]
That is my point, this answer is incomplete, the return code depends on you permissions:
404 - Only if you have `List`
403 - If you don't have `List`

**Comment by potiuk** [comment, marked as answer]
> IMO, It's not reasonable for us to swallow the permission error.

Fully agree with @xinbinhuang. The 404 is explicit "I have permissions, but the file does not exist". This is entirely different than "The file exist but I have no permissions". IMHO Replace=True does exactly what it should do - replaces stuff regardless if the file is there or whether you even have permission to list them. Replace = false should fail if you cannot check if the file is there - even if you have no list permission - generally if you have not enough permissions to check if the file is there, and if you want to use "replace = False", I'd expect to get exactly what you get - permission error (precisely because you have no permissions to check it). So for me this works as intended and expected. I will convert it into discussion if you want to further discuss it and close the PR, but I think it's just wrong to add 403 here.

**Reply by shlomi-viz** [reply]
I totally understand what you are saying, if `Replace=True` was the default it would make sense. 
In my mind I think, "Why should I care about replacing the file or not", when I ran the code and it failed, I had to read the source code to understand why, in my mind I only need `Put` permission to upload a file, so getting an error on `head_object` was very strange for me.
Lastly, I said it before, explicitly writing `Replace=True` only because I don't have permission to check if the file is already there feel too twisted.

Hope that make sense. 
Thanks

**Reply by potiuk** [reply]
So maybe you shoud simply update the doscstring stating the behaviour if there is no access? For me that's more than enough, and it's perfectly expected behaviour if replace is False by default (which I also think makes perfect sense and conscious choice).

Would you like to open PR clarifying this behaviour instead?

**Reply by xinbinhuang** [reply]
> Lastly, I said it before, explicitly writing `Replace=True` only because I don't have permission to check if the file is already there feel too twisted.


I would think of it this way instead: there are two ways of writing a file
- OVERWRITE/REPLACE approach: writing a file regardless if it already exists
- INSERT approach: upload a file only if it doesn't have a file

For the insert to work, you would need to check if the file already exists, so you need the `List` permission. If you do mean replace/overwrite, then you don't need to check.

> I totally understand what you are saying, if Replace=True was the default it would make sense.
> In my mind I think, "Why should I care about replacing the file or not", 
 
IMHO, I think you should care. Overwrite should be careful because it may cause data loss. If you accidentally overwrite the wrong file, you may not be able to recover it in the worst case. With this in mind, I would argue that dealing with a runtime exception is way better than accidental data loss.
